### PR TITLE
 [Enhance] Enhance ArcFaceClsHead

### PR DESCRIPTION
**Pull request by Ezra-Yu**
## Motivation

Enhance ArcFaceClsHead.

related paper : [subcenterArcFace](https://paperswithcode.com/paper/sub-center-arcface-boosting-face-recognition).
related issue #1169 

## Modification

0. refactor the Arcface Code
1. support subcenterArcFace
2. support AdvMargins
3. fix arcface when using amp

##### use vanilla ArcFace
```
mode = dict(
    backbone = xxx,
    neck = xxxx,
    head=dict(
        type='ArcFaceClsHead',
        num_classes=5000,
        in_channels=1024,
        loss = dict(type='CrossEntropyLoss', loss_weight=1.0),
        init_cfg=None),
)
```

##### use SubCenterArcFace with 3 sub-centers
```
mode = dict(
    backbone = xxx,
    neck = xxxx,
    head=dict(
        type='ArcFaceClsHead',
        num_classes=5000,
        in_channels=1024,
        num_subcenters=3,
        loss = dict(type='CrossEntropyLoss', loss_weight=1.0),
        init_cfg=None),
)
```

##### use ArcFaceWithAdvMargin

```
mode = dict(
    backbone = xxx,
    neck = xxxx,
    head=dict(
        type='ArcFaceClsHead',
        num_classes=5000,
        in_channels=1024,
        num_subcenters=3,
        loss = dict(type='CrossEntropyLoss', loss_weight=1.0),
        init_cfg=None),
)

custom_hooks = [dict(type='SetCountPowAdvMarginsHook')]
```

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by okotaku** [comment]
@Ezra-Yu The implementation is exactly as I intended in my proposal👍
It is a little complicated, but I can't think of any other simple plan.

File: mmcls/models/heads/arcface_head.py [hunk 1]
```diff
@@ -157,15 +266,16 @@ def loss(self, feats: Tuple[torch.Tensor],
...
             # Batch augmentation may convert labels to one-hot format scores.
             target = torch.stack([i.gt_label.score for i in data_samples])
         else:
-            target = torch.cat([i.gt_label.label for i in data_samples])
+            target = label_target
 
-        # The part can be traced by torch.fx
-        cls_score = self(feats, target)
+        # the index format target would be used
+        cls_score = self(feats, label_target)
```
Comment (okotaku) [inline mmcls/models/heads/arcface_head.py:278]: Implementation during mixup may not be correct.

I don't know which implementation is better, what do you think?
https://github.com/cavalleria/cavaface/blob/master/train.py
https://www.kaggle.com/competitions/happy-whale-and-dolphin/discussion/319941

Comment (Ezra-Yu) [reply mmcls/models/heads/arcface_head.py:278]: I think the first one, mixing the data, is more reasonable. Mixing the embedding seems weird. Just the embedding of the mixing, which I think only modifies the weight of the final average loss.

Support batch-augments is a little difficult in mmcls. Seem we can not get the specific target index and prem index in the head.

**Review by okotaku — APPROVED** [review]
LGTM

File: mmcls/models/heads/arcface_head.py (hunk 1 above)
Comment (Ezra-Yu) [reply mmcls/models/heads/arcface_head.py]: Now, Arcface should have supported train with batch-augments.  @okotaku

Comment (okotaku) [reply mmcls/models/heads/arcface_head.py]: It looks good.

Comment (Ezra-Yu) [reply mmcls/models/heads/arcface_head.py:284]: After discussion, considering to release a new version soon. We decided that in this PR, **only index-format targets are supported**, and for the batch-augment solution, we need to do some experiments and then determine.

The [first solution](https://github.com/cavalleria/cavaface/blob/master/train.py)  does not take into account the changes to logit by mix-target-label in https://github.com/cavalleria/cavaface/blob/822651f0e6d4d08df5441922acead39dc5375103/train.py#L261. The second one is just mix-embedding.

**Review by mzr1996 — APPROVED** [review]
